### ISSUE TITLE
Make nodeLabel available for templates.

### ISSUE THREAD

**Issue by velmeran** [OpenNMS/grafana-opennms-datasource#10]
When working with Templates there is no way to properly show the nodeLabel for each graph. Instead all we get is nodeID which is not human readable on large selections.  We need a way to do $var(label) or $var.label in the graphs Label section so that things will be more humanly readable.

**Comment by mhuot** [comment]
I'm wondering if this needs to be implemented as a function, similar to http://docs.grafana.org/datasources/graphite/

**Comment by deligatedgeek** [comment]
This would be really useful when showing multiple nodes on a single graph using templates.

Is there a solution to this?

**Comment by deligatedgeek** [comment]
I have found that setting the Foreign ID to the node ID is a temporary fix, but this defeats auto-discovery and dynamic dashboards in Grafana, as every device must be specified.

Current using the template variable name as Label shows Requisition Name:Foreign ID

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 05:21, deligatedgeek <notifications@github.com> wrote:
> 
> Current using the template variable name as Label shows Requisition Name:Foreign ID

On another note, why does it say requisition name in-lieu of foreign ID?

David

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 06:19, David Hustace <david@opennms.com> wrote:
> 
> On another note, why does it say requisition name in-lieu of foreign ID?

s/foreign ID/foreign source/

David

**Comment by deligatedgeek** [comment]
It displays Requisition Name:Foreign ID, I'm not sure why.  If I just add an interface without specifying Requisition Name then it only shows Foreign ID.

![grafanaopennmstemplatevariable](https://user-images.githubusercontent.com/22032477/29663402-d5188e36-88c2-11e7-8130-4d2e8a666b22.png)

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 06:53, deligatedgeek <notifications@github.com> wrote:
> 
> It displays Requisition Name:Foreign ID, I'm not sure why. If I just add an interface without specifying Requisition Name then it only shows Foreign ID.
> 
My question is more on the general note of the feature and not specific to your implementation.

Requisition Name isn't a property that we should be exposing. The property is foreignSource.

David

**Comment by mvrueden** [comment]
Just to clarify: Requisition name and foreign source are synonyms.

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 07:34, mvrueden <notifications@github.com> wrote:
> 
> Just to clarify: Requisition name and foreign source are synonyms.
> 
I realize that it's easy to make the connection but it really isn't and shouldn't be considered a synonym.

David

**Comment by mvrueden** [comment]
Yes, but that is not the case at the moment, that's why I pointed it out.

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 09:19, mvrueden <notifications@github.com> wrote:
> 
> Yes, but that is not the case at the moment, that's why I pointed it out.

I realize that, too, which is why I was pointing out the fallacy. :)

-David

**Comment by pbrane** [comment]
> On Aug 24, 2017, at 12:13, deligatedgeek <notifications@github.com> wrote:
> 
> The Gui now uses that terminology so I also used it here.
> I also used to manually setup foreign sources in a text file generated by a script, but I am trying to have a more dynamic configuration, so I decided to setup a Requisition with an discovery IP range

Again, I understand the confusion.  But the UI uses that terminology because it actually creates and manages requisitions.  So, in this case, it is true that the *foreign source* is the Requisition and, by convenience only, they share the same name even though they don't "have" to other than when used in the GUI.

However, the node's key (other than the internally generated database ID) is a composite of foreign source + foreign ID where the foreign source may or may not be a requisition.  The *source* may be a CMDB, DNS, or some custom API.  The design was to identify the *source* of the inventory that is being provisioned and use that as part of the composite key to not only identify the source but also to isolate the IDs to prevent collisions.

:D

**Comment by deligatedgeek** [comment]
Thanks for clarifying.
There was an update (#4) that listed node names in template variables on Grafana which was very useful, but when you select multiple nodes in the same graph there is a label conflict.  Using the template variable in the label presents the foreign source and ID.
The current solution is to have the template variable contain a predetermined list of nodes, which has to be manually updated.  This is a shame as the filtering in template variable is excellent and allows the selection based on almost any OpenNMS method of grouping, e.g. Catergory, has SNMP, has SMTP.
Therefore a dashboard can be created for a specific dynamic group of servers, they can be selected from a drop down and all selected nodes data displayed on each graph.  But the label would be a numeric string which users would have to look up by logging into OpenNMS.
Can the same fix be applied to graph labels as was applied to templating?
I know its actually not that simple as I would have just implemented it and submitted it for inclusion.
